inst/include/cpp11/protect.hpp

```diff
@@ -144,13 +131,23 @@ SEXP unwind_protect_sexp(Fun code) {
   if (setjmp(unwind_data.jmpbuf)) {
     throw unwind_exception(token);
   }
-  SEXP res = R_UnwindProtect(&internal::unwind_protect_unwrap_sexp<Fun>, &code,
-                             internal::maybe_jump, &unwind_data, token);
 
-  return res;
+  return R_UnwindProtect(
+      [](void* data) -> SEXP {
+        Fun* callback = (Fun*)data;
+        return (*callback)();
+      },
+      &code, internal::maybe_jump, &unwind_data, token);
 }
 
-template <typename Fun>
+template <typename Fun, typename = typename std::enable_if<
+                            std::is_same<decltype(code()), SEXP>::value>::type>
+SEXP unwind_protect(Fun code) {
+  return unwind_protect_sexp(code);
```

Comment:
@jimhester not sure if you'd prefer to remove or keep the explicitly suffixed overload

Reply:
If the general case below handles this case I think we can remove it.